aiostream/stream/combine.py

```diff
@@ -76,12 +78,17 @@ async def zip(*sources: AsyncIterable[T]) -> AsyncIterator[tuple[T, ...]]:
             await stack.enter_async_context(streamcontext(source)) for source in sources
         ]
         # Loop over items
+        STOP_SENTINEL = object()
         while True:
-            try:
-                coros = builtins.map(anext, streamers)
-                items = await asyncio.gather(*coros)
-            except StopAsyncIteration:
+            coros = (anext(streamer, STOP_SENTINEL) for streamer in streamers)
+            items = await asyncio.gather(*coros)
+            if all(item is STOP_SENTINEL for item in items):
                 break
+            elif any(item is STOP_SENTINEL for item in items):
+                if strict:
+                    raise ValueError("iterables have different lengths")
+                else:
+                    break
```

Comment:
@vxgmichel I wasn't sure why you still want to catch `StopAsyncIteration` at all for the non-strict case in your suggestion from https://github.com/vxgmichel/aiostream/issues/118#issuecomment-2217381371 instead of using the same logic of stopping based on a sentinel value for both strict and non-strict branches - here is my suggestion for how to do that.

But perhaps the `StopAsyncIteration`  has slightly better performance for the non-strict case? :thinking: Or another reason?

Reply:
> why you still want to catch StopAsyncIteration at all for the non-strict case in your suggestion

Here's my reasoning: imagine you zip a stream A with producing 3 elements really fast and a stream B producing 10 elements really slow. You will have to wait for stream B to produce the first three elements, but there no reason to wait for the fourth one if A is already exhausted and the `strict` parameter is `False`.

Reply:
I see, that makes sense :+1: I've added a test case for this (maybe you have a better idea for how to test it though) and changed it to use your approach.